buy: provide a first integration of Pocket widget

Beerosagos opened the pull request:
This commit is a first step in the integration process of Pocket widget into the buy workflow.

The widget is embedded in a dedicated iframe in `pocket.tsx`, while a drafted `exchange.tsx` allows the user to choose which exchange (moonpay or pocket) should be used in the purchase process. The iframe integration includes an address verification step similar to the one used with AOPP flow. In order to do this, a typescript `request-address` library has been added to the `package.json`.

Backend side, a new `pocket.go` contains the functions needed to verify coin support and iframe URL, while the code dedicated to the address verification is in `backend.go`.

benma commented on backend/exchanges/pocket.go:
```diff
@@ -0,0 +1,34 @@
+// BuyPocket verifies if the passed account is enabled for Pocket and returns the
+// url needed to incorporate the widget in the frontend.
+func BuyPocket(acct accounts.Interface) (string, error) {
+	if acct.Coin().Code() != coin.CodeBTC {
+		return "", fmt.Errorf("unsupported cryptocurrency code %q", acct.Coin().Code())
+	}
+
```
Use `IsPocketSupported()`?

is this check even needed? in the frontend this call doesn't happen unless you selected pocket (where the check already happened).

Beerosagos replied:
Right. Slightly modified the function to add TBTC support, as you proposed here https://github.com/digitalbitbox/bitbox-wallet-app/pull/1846#discussion_r1023795009

benma commented on backend/backend.go:
```diff
@@ -683,3 +685,62 @@ func (backend *Backend) HandleURI(uri string) {
+			Error("Pocket Widget: could not initialize account")
+		return "", ""
+	}
+	if account.Coin().Code() != coinpkg.CodeBTC {
```
use `IsPocketSupported()`?

Beerosagos replied:
ACK

benma commented on backend/exchanges/pocket.go:
```diff
@@ -0,0 +1,34 @@
+
+// IsPocketSupported is true if coin.Code is supported by Pocket.
+func IsPocketSupported(acct accounts.Interface) bool {
+	// Pocket would also support tbtc, but at the moment testnet address signing is disabled on firmware.
```
the actual check should then be (in pseudocode ;)):

(coinCode==BTC or coinCode==TBTC) and acct.keystore.CanSignMessage(coinCode)

then it will start working automatically if we enable tbtc msg signing in the bitbox02, or if we add signmsg to the software test keystore - the latter could actually be a good time now to test pocket integration in testnet.

Beerosagos replied:
Nice! Didn't notice `CanSignMessage` method :) Done :+1:

thisconnect commented on frontends/web/package.json:
```diff
@@ -20,7 +20,8 @@
     "react-i18next": "11.16.8",
-    "react-router-dom": "6.3.0"
+    "react-router-dom": "6.3.0",
+    "request-address": "^0.0.3"
```
Minor detail: for non-devdeps I would prefer to lock it in the end, but feel free to keep the semver range in staging.

To add a locked npm version do `npm install --save-exact request-address` or preferably `npm install --save-exact --ignore-scripts request-address`. The ignore-scripts flags does not execute postinstall hooks, there is almost never a reason to execute a script when installing npm deps.

Beerosagos replied:
Done :+1:

thisconnect commented on frontends/web/src/api/backend.ts:
```diff
@@ -77,3 +82,19 @@ export const getConfig = (): Promise<any> => {
+  return (): Promise<boolean> => {
+    return apiGet(`exchange/pocket/supported/${accountCode}`);
+  };
+};
```
maybe this is the right moment to start `api/exchanges.ts` and also move pocket-addr-sig in there?

Beerosagos replied:
ACK

thisconnect commented on frontends/web/src/routes/buy/exchange.tsx:
```diff
@@ -0,0 +1,61 @@
+  );
+};
+
+export const Exchange = translate()(ExchangeComponent);
```
Better use hooks instead of decorator, since this is a functional component

```ts
import { useTranslation } from 'react-i18next';

const ExchangeComponent: FunctionComponent<Props> = ({ code }) => {
  const { t } = useTranslation();
```

Beerosagos replied:
Removed until we don't decide which texts should be used and add the locales entry. There was the same mistake also in `pocket.tsx`, fixed there.

benma commented on backend/backend.go:
```diff
@@ -683,3 +685,62 @@ func (backend *Backend) HandleURI(uri string) {
+func (backend *Backend) PocketWidgetSignAddress(code accounts.Code, msg string, format string) (string, string) {
+	log := backend.log.WithField("accountCode", code)
+	var account accounts.Interface
+	for _, acct := range backend.accounts {
```
move `getAccountFromCode()` from the handlers here and use that in both places. That one also checks if the account is active.

Beerosagos replied:
Done :+1:

benma commented on frontends/web/src/api/backend.ts:
```diff
@@ -30,6 +30,11 @@ export interface ISuccess {
     errorCode?: string;
 }
 
+export type AddrSignResp = {
```
nit: we usually spell out all words, `AddressSignResponse`

Beerosagos replied:
ACK

benma commented on backend/handlers/handlers.go:
```diff
@@ -204,10 +205,13 @@ func NewHandlers(
 	getAPIRouter(apiRouter)("/aopp/cancel", handlers.postAOPPCancelHandler).Methods("POST")
 	getAPIRouter(apiRouter)("/aopp/approve", handlers.postAOPPApproveHandler).Methods("POST")
 	getAPIRouter(apiRouter)("/aopp/choose-account", handlers.postAOPPChooseAccountHandler).Methods("POST")
+	getAPIRouter(apiRouter)("/pocket-addr-sign", handlers.postPocketWidgetSignAddress).Methods("POST")
```
nit: we usually spell out words: address not addr.

super-nit: `address-sign` vs `SignAddress` - why the different order of the words? :smile: 

could also do `/pocket/...` in case more endpoints should come.

Beerosagos replied:
Ok! about the super-nit: I think I changed it in a second moment and didn't notice the discrepancy :D

benma commented on backend/backend.go:
```diff
@@ -683,3 +685,62 @@ func (backend *Backend) HandleURI(uri string) {
 }
+
+// PocketWidgetSignAddress returns an unused address and makes the user sign a message to prove ownership.
+func (backend *Backend) PocketWidgetSignAddress(code accounts.Code, msg string, format string) (string, string) {
```
return (string, string, error) and return all errors and propagate them all the way to the frontend, which should be able to react/show failure modes, if the user aborted, etc.

Beerosagos replied:
Right, done!

benma commented on backend/backend.go:
```diff
@@ -683,3 +685,62 @@ func (backend *Backend) HandleURI(uri string) {
 	}
 }
+
+// PocketWidgetSignAddress returns an unused address and makes the user sign a message to prove ownership.
```
document both return values more clearly, especially that the sig is returned in base64

Beerosagos replied:
Done, hope it is fine :)

benma commented on frontends/web/src/routes/buy/exchange.tsx:
```diff
@@ -0,0 +1,61 @@
+ */
+
+import { translate, TranslateProps } from '../../decorators/translate';
+import { FunctionComponent } from 'react';
```
we don't use that anymore afaik, see https://github.com/digitalbitbox/bitbox-wallet-app/pull/1839

Beerosagos replied:
Nice! Fixed

benma commented on backend/handlers/handlers.go:
```diff
@@ -1068,6 +1072,34 @@ func (handlers *Handlers) getExchangeMoonpayBuy(r *http.Request) (interface{}, e
+		return nil, err
+	}
+
+	return exchanges.BuyPocket(acct)
```
fyi, errors returned like this are swallowed in Qt (we should refactor error handling in the handlers). if you want to handle errors, for now we normally return a struct with a success flag and error message.

Beerosagos replied:
Fixed

benma commented on frontends/web/src/routes/buy/pocket.tsx:
```diff
@@ -0,0 +1,144 @@
+          <Header title={<h2>{t('buy.info.title', { name })}</h2>} />
+        </div>
+        <div ref={ref} className="innerContainer">
+          <div className="noSpace" style={{ height }}>
```
slightly confused, what is double braces? and isn't `height` an int not a style?

benma replied:
TIL!

benma commented on frontends/web/src/routes/buy/pocket.tsx:
```diff
@@ -0,0 +1,144 @@
+          .then( resp => {
+            console.log(JSON.stringify(resp));
+
+            if (!resp.addr || !resp.sig ) {
```
@thisconnect wasn't there a linter for spaces? also above before `resp`

thisconnect replied:
made a PR https://github.com/digitalbitbox/bitbox-wallet-app/pull/1847

Beerosagos replied:
spaces fixed, in the meantime

benma commented on backend/backend.go:
```diff
@@ -683,3 +685,62 @@ func (backend *Backend) HandleURI(uri string) {
+		account.Config().SigningConfigurations[signingConfigIdx].ScriptType(),
+	)
+	if err != nil {
+		if firmware.IsErrorAbort(err) {
```
this would definitely be helpful to propagate. maybe an 'user aborted' flag/msg should even be accepted by pocket? we should ask them about it.

Beerosagos replied:
Added an `abort` flag in the `Response` struct. Atm the frontend just shows an alert, but we could send a message to the widget for this specific case, if we decide so.

Beerosagos commented:
Converted to draft until I finish addressing all the comments, I'll be pushing partial improvements in the meantime :pray:

Beerosagos commented on frontends/web/src/routes/buy/exchange.tsx:
```diff
@@ -0,0 +1,61 @@
+        { showPocket && (<Button
+          primary
+          onClick={() => goToExchange('pocket')} >
+          {t('Pocket')}
```
Sorry, was just a placeholder :) Removed

Beerosagos commented on frontends/web/src/routes/buy/exchange.tsx:
```diff
@@ -0,0 +1,61 @@
+
+  return (
+    <div>
+      <div>{t('Choose you exchange!')}</div>
```
Also here, just a placeholder. Removed until we don't define layout and texts of the page

Beerosagos commented on backend/exchanges/pocket.go:
```diff
@@ -0,0 +1,88 @@
+// 	`message` is the message that will be signed by the user with the private key linked to the address.
+//	`format` is the script type that should be used in the address derivation, as received by the widget
+//		(see https://github.com/pocketbitcoin/request-address#requestaddressv0messagescripttype).
+// 	`aoppBTCScriptTypeMap` is the map used in the AOPP flow to get the `ScriptType` object related to the `format` param.
```
Not sure if this is totally correct. @benma do you think we should make a dedicated struct or see any other better solution?

Beerosagos commented:
Thanks @thisconnect and @benma for the review. Took some time to address the comment and to refactor the backend. 

I modified most of the function names, hope they are more intuitive now. I also moved `PocketWidgetSignAddress` in `pocket.go`, so that now all the logic should be in the same file. 

Could you take another look when you have a moment?

thisconnect commented on frontends/web/src/routes/buy/pocket.tsx:
```diff
@@ -0,0 +1,151 @@
+      return;
+    }
+
+    const iframeOrigin = iframeURL.match(/^https?:\/\/[^/?]+/)![0];
```
Maybe change to `new URL(location.href).origin`. If the url can be invalid it should be in a try/catch, but I think in this case it is fine.

Beerosagos replied:
Ah nice! Thank you :) Now should be better

benma reviewed (approved):
LGTM, awesome

thisconnect reviewed (approved):
LGTM
